src/main/java/org/pac4j/jax/rs/resteasy/Pac4JProfileInjectorFactory.java

```diff
@@ -0,0 +1,79 @@
+package org.pac4j.jax.rs.resteasy;
+
+import org.jboss.resteasy.core.InjectorFactoryImpl;
+import org.jboss.resteasy.core.ValueInjector;
+import org.jboss.resteasy.spi.HttpRequest;
+import org.jboss.resteasy.spi.HttpResponse;
+import org.jboss.resteasy.spi.ResteasyProviderFactory;
+import org.jboss.resteasy.spi.metadata.Parameter;
+import org.jboss.resteasy.util.FindAnnotation;
+import org.pac4j.core.context.J2EContext;
+import org.pac4j.core.profile.ProfileManager;
+import org.pac4j.jax.rs.annotations.Pac4JProfile;
+import org.pac4j.jax.rs.annotations.Pac4JProfileManager;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.AccessibleObject;
+import java.lang.reflect.Type;
+import java.util.Optional;
+import java.util.function.BiFunction;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.ws.rs.ext.Provider;
+
+@Provider
+public class Pac4JProfileInjectorFactory extends InjectorFactoryImpl {
+
+	@Override
+	public ValueInjector createParameterExtractor(Class injectTargetClass, AccessibleObject injectTarget, Class type,
+												  Type genericType, Annotation[] annotations, ResteasyProviderFactory factory) {
+		final ValueInjector injector = getValueInjector(type, annotations);
+		if (injector != null) return injector;
+		return super.createParameterExtractor(injectTargetClass, injectTarget, type, genericType, annotations, factory);
+	}
+
+	@Override
+	public ValueInjector createParameterExtractor(Parameter parameter, ResteasyProviderFactory providerFactory) {
+		final ValueInjector injector = getValueInjector(parameter.getType(), parameter.getAnnotations());
+		if (injector != null) return injector;
+		return super.createParameterExtractor(parameter, providerFactory);
+	}
+
+	private ValueInjector getValueInjector(Class type, Annotation[] annotations) {
+		final Pac4JProfile profile;
+		if ((profile = FindAnnotation.findAnnotation(annotations, Pac4JProfile.class)) != null) {
+			if (type.equals(Optional.class)) {
+				return new Pac4JValueInjector(
+						(req, resp) -> new ProfileManager(new J2EContext(req, resp)).get(profile.readFromSession())
+				);
+			} else {
+				return new Pac4JValueInjector(
+						(req, resp) -> new ProfileManager(new J2EContext(req, resp)).get(profile.readFromSession()).orElse(null)
```

Comment:
In the jersey implementation, I preferred to throw an `new WebApplicationException(401)` instead of returning null.
The rational is that if you want to retrieve a profile that can be potentially not there (i.e. you are not authenticated), then you use the Optional.
If not (i.e., you are expecting to be authenticated, then the profile shouldn't be null, then we should't allow for such situation, especially because it introduces a security risks.

Actually one of the test relies on this behaviour ;)

See https://github.com/pac4j/jax-rs-pac4j/blob/master/src/main/java/org/pac4j/jax/rs/jersey/features/Pac4JValueFactoryProvider.java#L246